lxd/util/net.go

```diff
@@ -132,3 +132,36 @@ func NetworkInterfaceAddress() string {
 	}
 	return ""
 }
+
+// IsAddressCovered detects if network address1 is actually covered by
+// address2, in the sense that they are either the same address or address2 is
+// specified using a wildcard with the same port of address1.
+func IsAddressCovered(address1, address2 string) bool {
+	if address1 == address2 {
+		return true
+	}
+
+	_, port1, err := net.SplitHostPort(address1)
+	if err != nil {
+		return false
+	}
+
+	host2, port2, err := net.SplitHostPort(address2)
+	if err != nil {
+		return false
+	}
+
+	// If the ports are different, then address1 is clearly not covered by
+	// address2.
+	if port2 != port1 {
+		return false
+	}
+
+	// If the host part of the address2 is a wildcard, then address1 is
+	// covered, since it has the same port.
+	if host2 == "0.0.0.0" || host2 == "::" {
```

Comment:
Almost but not quite true. `[::]` will bind all IPv6 and IPv4 address, `0.0.0.0` will bind all IPv4 addresses only.
So if I have `[::1]:8443` and `0.0.0.0:8443`, the latter does not in fact cover the former.

Adding a check for host2 being `0.0.0.0` and then checking if host1 is IPv6 should do the trick, if it is IPv6, then it's not in fact covered.

I'm also unsure how this function will behave if I pass `:8443` as the address, this should normally be treated as the same as `[::]:8443` I think.

Reply:
Nice catch. I didn't know about this nuance. I've changed the code as you suggested, adding logic to handle the ```:8443``` case as well (which is equivalent to ```[::]:8443```, as you expected).

You can see the behavior of the function in the various cases in this [test](https://github.com/freeekanayaka/lxd/blob/cluster-https-address-support/lxd/util/net_test.go#L54).